includes/ProductSync/ProductValidator.php

```diff
@@ -0,0 +1,203 @@
+use WC_Product;
+use WC_Facebookcommerce_Integration;
+
+/**
+ * Class ProductValidator
+ *
+ * This class is responsible for validating whether a product should be synced to Facebook.
+ *
+ * @since 2.5.0
+ */
+class ProductValidator {
+
+	/**
+	 * @var string the meta key used to flag whether a product should be synced in Facebook
+	 */
+	const SYNC_ENABLED_META_KEY = '_wc_facebook_sync_enabled';
+
+	/**
+	 * @var WC_Facebookcommerce_Integration
+	 */
+	protected $integration;
+
+	/**
+	 * @var WC_Product The product object to validate.
+	 */
+	protected $product;
+
+	/**
+	 * @var WC_Product The product parent object if the product has a parent.
+	 */
+	protected $product_parent;
+
+	/**
+	 * ProductValidator constructor.
+	 *
+	 * @param WC_Facebookcommerce_Integration $integration
+	 * @param WC_Product                      $product The product to validate. Accepts both variations and variable products.
+	 */
+	public function __construct( WC_Facebookcommerce_Integration $integration, WC_Product $product ) {
+		$this->product = $product;
+
+		if ( $product->get_parent_id() ) {
+			$parent_product = wc_get_product( $product->get_parent_id() );
+			if ( $parent_product instanceof WC_Product ) {
+				$this->product_parent = $parent_product;
+			}
+		}
+
+		$this->integration = $integration;
+	}
+
+	/**
+	 * Validate whether a given product should be synced to Facebook.
+	 *
+	 * @throws ProductExcludedException If product should not be synced.
+	 */
+	public function validate() {
+		$this->validate_sync_enabled_globally();
+		$this->validate_product_status();
+		$this->validate_product_stock_status();
+		$this->validate_product_sync_field();
+		$this->validate_product_price();
+		$this->validate_product_visibility();
+		$this->validate_product_categories_and_tags();
+	}
+
+	/**
+	 * Validate whether a given product should be synced to Facebook but skip the status check for backwards compatibility.
+	 *
+	 * @internal Do not use this as it will likely be removed.
+	 *
+	 * @throws ProductExcludedException If product should not be synced.
+	 */
+	public function validate_but_skip_status_check() {
+		$this->validate_sync_enabled_globally();
+		$this->validate_product_stock_status();
+		$this->validate_product_sync_field();
+		$this->validate_product_price();
+		$this->validate_product_visibility();
+		$this->validate_product_categories_and_tags();
+	}
+
+	/**
+	 * Check whether product sync is globally disabled.
+	 *
+	 * @throws ProductExcludedException If product should not be synced.
+	 */
+	protected function validate_sync_enabled_globally() {
+		if ( $this->integration->is_product_sync_enabled() ) {
+			throw new ProductExcludedException( 'Product sync is globally disabled.' );
+		}
+	}
+
+	/**
+	 * Check whether the product's status excludes it from sync.
+	 *
+	 * @throws ProductExcludedException If product should not be synced.
+	 */
+	protected function validate_product_status() {
+		$product = $this->product_parent ?: $this->product;
```

Comment:
Using short ternaries is not allowed (WordPress.PHP.DisallowShortTernary.Found)
I wonder why this is was not annotated.

Reply:
> Personally, I think we should globally allow short ternaries as they are very useful.
Although I agree about the usefulness, if we want to change this we should convince the WP community to this change. Keeping standards helps us interact better with the outside world. https://make.wordpress.org/core/2019/07/12/php-coding-standards-changes/